.ci/magician/cmd/scheduled_pr_reminders.go

```diff
@@ -0,0 +1,419 @@
+// The basic algorithm is:
+// - find the most recent request for review from a core contributor
+//   - if there are none, the state is waitingForReviewerAssignment
+//
+// - check for any reviews from core reviewers since that review request.
+//   - if there are none, the state is waitingForReview and the time is the
+//     review request time
+//   - if any are change requests, the state is waitingForContributor and the time
+//     is the earliest change request
+//   - if any are approvals, the state is waitingForMerge and the time is the
+//     earliest approval
+//   - otherwise there are reviews and all are comment reviews; the state is
+//     waitingForContributor and the time is the earliest review time
+//
+// We don't specially handle cases where the contributor has "acted" because it would be
+// significant additional effort, and this case is already handled by re-requesting review
+// automatically based on contributor actions.
+func notificationState(pr *github.PullRequest, issueEvents []*github.IssueEvent, reviews []*github.PullRequestReview) (pullRequestReviewState, time.Time, error) {
+	slices.SortFunc(issueEvents, func(a, b *github.IssueEvent) int {
+		if a.CreatedAt.Before(*b.CreatedAt.GetTime()) {
+			return 1
+		}
+		if a.CreatedAt.After(*b.CreatedAt.GetTime()) {
+			return -1
+		}
+		return 0
+	})
+	slices.SortFunc(reviews, func(a, b *github.PullRequestReview) int {
+		if a.SubmittedAt.Before(*b.SubmittedAt.GetTime()) {
+			return 1
+		}
+		if a.SubmittedAt.After(*b.SubmittedAt.GetTime()) {
+			return -1
+		}
+		return 0
+	})
+
+	var latestReviewRequest *github.IssueEvent
+	for _, event := range issueEvents {
+		if *event.Event != "review_requested" {
+			continue
+		}
+		if event.RequestedReviewer == nil {
+			continue
+		}
+		if membership.IsCoreReviewer(*event.RequestedReviewer.Login) {
+			latestReviewRequest = event
+			break
+		}
+	}
+
+	if latestReviewRequest == nil {
+		return waitingForReviewerAssignment, *pr.CreatedAt.GetTime(), nil
+	}
+
+	var earliestApproved *github.PullRequestReview
+	var earliestChangesRequested *github.PullRequestReview
+	var earliestCommented *github.PullRequestReview
+
+	seenReviewers := map[string]struct{}{}
+	approvedBy := map[string]struct{}{}
+	for _, review := range reviews {
+		if review.SubmittedAt.Before(*latestReviewRequest.CreatedAt.GetTime()) {
+			break
+		}
+		// Ignore reviews by deleted accounts
+		if review.User == nil {
+			continue
+		}
+		if !membership.IsCoreReviewer(*review.User.Login) {
+			continue
+		}
+		reviewer := *review.User.Login
+
+		// ignore any reviews by reviewers who had a later approval
+		if _, ok := approvedBy[reviewer]; ok {
+			continue
+		}
+		switch *review.State {
+		case "DISMISSED":
+			// ignore dismissed reviews
+			continue
+		case "APPROVED":
+			approvedBy[reviewer] = struct{}{}
+			// only consider approvals that are the most recent review by the reviewer.
```

Comment:
I think the overall system looks good to me. There's only one specific scenario that I'm not certain about - comments after approval from the same contributor. I wonder if we should consider sending reminders to reviewers instead of contributors in such cases, as if the following comment should block the PR merge, the reviewer should probably dismiss the previous approval instead. One common reason that I can think of for this case is reviewer leave comment to announce pause on merging PRs due to probably CI issues.

Reply:
A new review of type "COMMENTED" would dismiss a previous review of type "APPROVED" from the same contributor. New comments that are not part of a review would have no impact.

I testing this on https://github.com/GoogleCloudPlatform/magic-modules/pull/10495 and GitHub's approval behavior is:

1. I approve. State: approved.
2. I add a COMMENTED review. State: approved.
3. I add a CHANGES_REQUESTED review. State: Changes requested.
4. I dismiss the CHANGES_REQUESTED review. State: Unreviewed.

This end state has 3 reviews which are APPROVED, COMMENTED, and DISMISSED: https://api.github.com/repos/GoogleCloudPlatform/magic-modules/pulls/10495/reviews

This behavior seems clearly inconsistent to me; the dismissed review should be ignored and return the PR to its approved state.

After thinking on it some more, I am ambivalent about this; making COMMENTED not override APPROVED seems like a reasonable compromise.

Reply:
Changed in a86c375ac